atlasdb-client/src/main/java/com/palantir/atlasdb/util/SlidingWindowMetricsInvocationHandler.java

```diff
@@ -0,0 +1,100 @@
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.SlidingTimeWindowArrayReservoir;
+import com.codahale.metrics.Timer;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.tritium.event.AbstractInvocationEventHandler;
+import com.palantir.tritium.event.DefaultInvocationContext;
+import com.palantir.tritium.event.InstrumentationProperties;
+import com.palantir.tritium.event.InvocationContext;
+
+/**
+ * A modified version of {@link com.palantir.tritium.event.metrics.MetricsInvocationEventHandler} that uses a
+ * {@link SlidingTimeWindowArrayReservoir} for the timer's reservoir.
+ */
+public final class SlidingWindowMetricsInvocationHandler extends AbstractInvocationEventHandler<InvocationContext> {
+    private static final Logger logger = LoggerFactory.getLogger(SlidingWindowMetricsInvocationHandler.class);
+    private static final String FAILURES_METRIC_NAME = "failures";
+
+    private final MetricRegistry metricRegistry;
+    private final String serviceName;
+
+    public SlidingWindowMetricsInvocationHandler(MetricRegistry metricRegistry, String serviceName) {
+        super(InstrumentationProperties.getSystemPropertySupplier(serviceName));
+        this.metricRegistry = checkNotNull(metricRegistry, "metricRegistry");
+        this.serviceName = checkNotNull(serviceName, "serviceName");
+    }
+
+    @Override
+    public InvocationContext preInvocation(Object instance, Method method, Object[] args) {
+        return DefaultInvocationContext.of(instance, method, args);
+    }
+
+    @Override
+    public void onSuccess(@Nullable InvocationContext context, @Nullable Object result) {
+        if (context == null) {
+            logger.debug("Encountered null metric context likely due to exception in preInvocation");
+            return;
+        }
+
+        long nanos = System.nanoTime() - context.getStartTimeNanos();
+        metricRegistry.timer(getBaseMetricName(context), SlidingWindowMetricsInvocationHandler::createNewTimer)
+                .update(nanos, TimeUnit.NANOSECONDS);
+    }
+
+    @Override
+    public void onFailure(@Nullable InvocationContext context, @Nonnull Throwable cause) {
+        if (context == null) {
+            markGlobalFailure();
+            logger.debug("Encountered null metric context likely due to exception in preInvocation: {}",
+                    UnsafeArg.of("cause", cause),
+                    cause);
+            return;
+        }
+
+        markGlobalFailure();
+        String failuresMetricName = MetricRegistry.name(getBaseMetricName(context), FAILURES_METRIC_NAME);
+        metricRegistry.meter(failuresMetricName).mark();
+        metricRegistry.meter(MetricRegistry.name(failuresMetricName, cause.getClass().getName())).mark();
+    }
+
+    private static Timer createNewTimer() {
+        return new Timer(new SlidingTimeWindowArrayReservoir(35, TimeUnit.SECONDS));
```

Comment:
curious why 35 second sliding window?

Reply:
30 sec is our collection rate I believe, and didn't want to have anything missed out due to weird timing things, so put it at 35. Ideally we'd want it at 30 - to be fair, I'm not sure I tried with 30 seconds whilst testing.

Reply:
also re window size, if we keep it small, we have to worry less about memory footprint since we're only keeping last 30 sec of metrics in memory as opposed to last 5 mins for example plus dd stops lying to me